### PR TITLE
Right-click info

### LINKED ISSUE #205: Interrogation clic-droit
Avec le géoportail actuel il est possible de faire un clic-droit pour obtenir un certain nombre d'informations relatives à la position géographique cliquée.

Voici à quoi ressemble la popup de résultats :

right-click-popup

La popup contient donc :

* des permaliens
* un QR code (permalien)
* des liens liés au routing
* la position du point selon 4 systèmes de projection

Questions :

* conserve-t-on cette interrogation/popup clic-droit ?
* quelles informations y mettons-nous ?

### PR DESCRIPTION
**Pull request by petzlux**
closes #205

### CODE REVIEW

File: geoportailv3/static/js/shorturlservice.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+ * @param {angular.$http} $http The Angular $http service.
+ * @param {ngeo.Location} ngeoLocation The ngeo Location service.
+ * @param {string} shorturlServiceUrl The URL to the "shorturl" service.
+ * @return {app.GetShorturl} The getShorturl function.
+ * @private
+ * @ngInject
+ */
+app.getShorturl_ = function($http, ngeoLocation, shorturlServiceUrl) {
+  return (
+      /**
```
Comment (elemoine) [inline geoportailv3/static/js/shorturlservice.js]: Indent issue.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+app.module.directive('appLocationinfo', app.locationinfoDirective);
+
+
+
+/**
+ * @constructor
+ * @param {angular.$http} $http
+ * @param {angular.Scope} $scope
+ * @param {app.GetShorturl} appGetShorturl
+ * @param {string} appLocationinfoTemplateUrl
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: `@inject` require here!

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+   * @type {app.GetShorturl}
+   * @private
+   */
+  this.getShorturl_ = appGetShorturl;
+
+  map.getViewport().addEventListener('contextmenu', goog.bind(function(event) {
+    event.preventDefault();
+    this['infoOpen'] = true;
+    var clickCoordinate = this.map.getEventCoordinate(event);
+    var feature = /** @type {ol.Feature} */
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: You don't need a type-cast here.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        (new ol.Feature(new ol.geom.Point(clickCoordinate)));
+    var features = this.featureOverlay_.getFeatures();
+    features.clear();
+    features.push(feature);
+    this.getShorturl_(clickCoordinate).then(goog.bind(
+        /**
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: Indent issue.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+   * @private
+   */
+  this.featureOverlay_ = this.createFeatureOverlay_();
+
+  $scope.$watch(goog.bind(function() {
+    return this['infoOpen'];
+  }, this), goog.bind(function(newVal) {
+    if (newVal === false) {
+      var features = this.featureOverlay_.getFeatures();
+      features.clear();
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: For performance and robustness reasons I think it'd be best to remove the feature overlay from the map (using `this.featureOverlay_.setMap(null)`.

File: geoportailv3/static/js/shorturlservice.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+     * @param {ol.Coordinate=} opt_coordinate
+     * @return {!angular.$q.Promise} Promise providing the short URL.
+     */
+      function getShorturl(opt_coordinate) {
+        var location = ngeoLocation;
+        if (opt_coordinate) {
```
Comment (elemoine) [inline geoportailv3/static/js/shorturlservice.js]: Please use `goog.isDef`.

File: geoportailv3/static/js/shorturlservice.js [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+ * @ngInject
+ */
+app.getShorturl_ = function($http, ngeoLocation, shorturlServiceUrl) {
+  return (
+      /**
+     * @param {ol.Coordinate=} opt_coordinate
+     * @return {!angular.$q.Promise} Promise providing the short URL.
+     */
+      function getShorturl(opt_coordinate) {
+        var location = ngeoLocation;
```
Comment (elemoine) [inline geoportailv3/static/js/shorturlservice.js]: This local variable is not really useful.

File: geoportailv3/static/js/shorturlservice.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            'Y': Math.round(opt_coordinate[1])
+          });
+        }
+        var url = location.getUriString();
+        return $http.get(shorturlServiceUrl, {
+          params: {
+            'url': url
+          }
+        }).then(
+            /**
```
Comment (elemoine) [inline geoportailv3/static/js/shorturlservice.js]: Indent issue.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js (hunk 2 above)
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: I think the non-debug mode should work if you add the `@ngInject` annotation.

**Comment by elemoine** [comment]
Done reviewing. I think the code does not work in non-debug mode because of the missing `@ngInject` annotation for the controller constructor.

**Comment by petzlux** [comment]
Arghh, your right, missing @ngInject , did not see the tree for the forest ! Merci !

File: geoportailv3/static/js/shorturlservice.js (hunk 1 above)
Comment (petzlux) [inline geoportailv3/static/js/shorturlservice.js]: fixjsstyle creates this indent ... is this a bug ?

File: geoportailv3/static/js/shorturlservice.js (hunk 8 above)
Comment (petzlux) [inline geoportailv3/static/js/shorturlservice.js]: fixjsstyle creates this indent, seems like a bug

Comment (elemoine) [inline geoportailv3/static/js/shorturlservice.js]: I've had bad experience with fixjsstyle and I never use it.

On Tue, Jun 2, 2015 at 4:20 PM, Patrick Weber <notifications@github.com>
wrote:

> In geoportailv3/static/js/shorturlservice.js
> <https://github.com/Geoportail-Luxembourg/geoportailv3/pull/507#discussion_r31527097>
> :
>
> > +     */
> > +      function getShorturl(opt_coordinate) {
> > +        var location = ngeoLocation;
> > +        if (opt_coordinate) {
> > +          location.updateParams({
> > +            'X': Math.round(opt_coordinate[0]),
> > +            'Y': Math.round(opt_coordinate[1])
> > +          });
> > +        }
> > +        var url = location.getUriString();
> > +        return $http.get(shorturlServiceUrl, {
> > +          params: {
> > +            'url': url
> > +          }
> > +        }).then(
> > +            /**
>
> fixjsstyle creates this indent, seems like a bug
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Geoportail-Luxembourg/geoportailv3/pull/507/files#r31527097>
> .
>



-- 
Eric Lemoine

Camptocamp France SAS
Savoie Technolac, BP 352
73377 Le Bourget du Lac, Cedex

Tel : +33 (0)4 58 48 20 34
Mail : eric.lemoine@camptocamp.com
http://www.camptocamp.com

**Comment by petzlux** [comment]
todo: elevation service

**Comment by petzlux** [comment]
Ready to merge?

**Comment by jaykayone** [comment]
let me do user validation first

**Comment by jaykayone** [comment]
* WGS 84: make two strings, one for dms, one for ll
* Top align "Short Url" and qrcode
* Bus links
* reverse geocoding

**Comment by petzlux** [comment]
Question for @elemoine : How can we implement the right click on touch devices ? I have looked if ol3 has long press, but couldnt find anything

**Comment by elemoine** [comment]
> Question for @elemoine : How can we implement the right click on touch devices ? I have looked if ol3 has long press, but couldnt find anything

No ol3 does not provide anything related to that. mf-geoadmin3 may be a good place to look at.

**Comment by petzlux** [comment]
Question for @pgiraud & @elemoine 

We have an issue trying to integrate layer query and location info.

On mobile touch devices, we want short tap to show layer query, and long tap (300ms) for location info. Both work, but there is an undesired interaction between the two.

Specifically, a long tap raises also a singleclick event, which means that layer query results are always shown in the info window, even when the user only wanted location information. 

How can we distinguish between short and long taps ?

**Comment by petzlux** [comment]
We resolved the long and short tap issue in https://github.com/Geoportail-Luxembourg/geoportailv3/commit/6747e6a4f3f6ca16ee88ab884afc46eda77633ae

**Comment by petzlux** [comment]
Ready for review ...

**Comment by petzlux** [comment]
@elemoine Can we get this reviewed and merged ?

File: geoportailv3/static/js/elevationservice.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    var lonlat = /** @type {ol.Coordinate} */
+        (ol.proj.transform(coordinate,
+            'EPSG:3857', 'EPSG:2169'));
+    return $http.get(elevationServiceUrl, {
+      params: {
+        'lon': lonlat[0],
+        'lat': lonlat[1]
+      }
+    }).then(
+        /**
```
Comment (elemoine) [inline geoportailv3/static/js/elevationservice.js]: Indent issue.

File: geoportailv3/static/js/elevationservice.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+           * @param {angular.$http.Response} resp Ajax response.
+           * @return {string} The elevation
+           */
+            function(resp) {
+              if (resp.data['dhm'] > 0) {
+                return parseInt(resp.data['dhm'] / 100, 0).toString() + ' m';
+              } else {
+                return 'N/A';
```
Comment (elemoine) [inline geoportailv3/static/js/elevationservice.js]: No translation needed for that?

File: geoportailv3/static/js/infobar/elevationdirective.js [hunk 11]
```diff
@@ -46,26 +47,33 @@ app.module.directive('appElevation', app.elevationDirective);
  * @ngInject
  * @constructor
  * @param {angular.$http} $http
+ * @param {app.GetElevation} appGetElevation
  * @param {ngeo.Debounce} ngeoDebounce
- * @param {string} elevationServiceUrl
  */
 app.ElevationDirectiveController =
-    function($http, ngeoDebounce, elevationServiceUrl) {
+    function($http, appGetElevation, ngeoDebounce) {
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/elevationdirective.js]: Very minor but we usually place ngeo services before app services in the function arguments.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+   */
+  this.startPixel_ = null;
+
+  /**
+   * @type {angular.$q.Promise|undefined}
+   * @private
+   */
+  this.holdPromise_ = undefined;
+
+  this['map'].on('pointerdown', goog.bind(function(event) {
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: Please use `goog.events.listen` as done elsewhere. In this case:

```js

goog.require('goog.events');
goog.require('ol.MapBrowserEvent.EventType');

goog.events.listen(map, ol.MapBrowserEvent.EventType.POINTERDOWN, function(event) {
  // …
}, false, this);
```

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+      }, this));
+
+  /**
+   * @type {angular.$timeout}
+   * @private
+   */
+  this.$timeout_ = $timeout;
+
+  /**
+   * @type {?Object<number, number>}
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: I think don't need the question mark here, as objects are nullable by default.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 14]
```diff
@@ -0,0 +1,278 @@
...
+  this.getShorturl_ = appGetShorturl;
+
+  /**
+   * @type {string}
+   * @private
+   */
+  this.qrServiceUrl_ = qrServiceUrl;
+
+  /**
+   * @type {Object}
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: You could use `Object.<string, string>`.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+   * @private
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: You could use `@const` to indicate to the compiler that this should not change.

File: geoportailv3/static/js/maincontroller.js [hunk 16]
```diff
@@ -127,6 +127,11 @@ app.MainController = function($scope, ngeoGetBrowserLanguage, gettextCatalog,
   this['userOpen'] = false;
 
   /**
+   * @type {string|undefined}
+   */
+  this['infosAppSelector'] = undefined;
```
Comment (elemoine) [inline geoportailv3/static/js/maincontroller.js]: What is this for? What are the possible values? Could this be a `boolean`?

File: geoportailv3/static/js/elevationservice.js [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+app.getElevation_ = function($http, elevationServiceUrl) {
+  return getElevation;
+
+  /**
+       * @param {ol.Coordinate} coordinate
+       * @return {!angular.$q.Promise} Promise providing the short URL.
+       */
+  function getElevation(coordinate) {
+    var lonlat = /** @type {ol.Coordinate} */
+        (ol.proj.transform(coordinate,
```
Comment (elemoine) [inline geoportailv3/static/js/elevationservice.js]: You need a `goog.require` for `ol.proj`.

File: geoportailv3/static/js/coordinatestringservice.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+    coordinate = ol.proj.transform(coordinate, sourceEpsgCode, targetEpsgCode);
+
+    switch (targetEpsgCode) {
+      case 'EPSG:2169':
+        str = ol.coordinate.format(coordinate, '{x} E | {y} N', 0);
+        break;
+      case 'EPSG:4326':
+        if (goog.isDef(opt_DMS) && opt_DMS === true) {
+          var hdms = ol.coordinate.toStringHDMS(coordinate);
```
Comment (elemoine) [inline geoportailv3/static/js/coordinatestringservice.js]: You need `goog.require` for `ol.coordinate`.

File: geoportailv3/static/js/infobar/projectionselectordirective.js [hunk 19]
```diff
@@ -48,76 +49,51 @@ app.module.directive('appProjectionselector', app.projectionselectorDirective);
...
-      break;
-    case 'EPSG:32631':
-      str = ol.coordinate.format(coord, '{x} | {y} (UTM31N)', 0);
-      break;
+app.ProjectionselectorDirectiveController.prototype.mouseCoordinateFormat_ =
+    function(coord) {
+  var mapEpsgCode =
+      this['map'].getView().getProjection().getCode();
+  if (this['projection']['value'] === 'EPSG:4326:DMS') {
+    var epsgCode = goog.string.remove(this['projection']['value'], ':DMS');
```
Comment (elemoine) [inline geoportailv3/static/js/infobar/projectionselectordirective.js]: `goog.require('goog.string')`

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+  this.$timeout_ = $timeout;
+
+  /**
+   * @type {?Object<number, number>}
+   * @private
+   */
+  this.startPixel_ = null;
+
+  /**
+   * @type {angular.$q.Promise|undefined}
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: I'd rather use `{?angular.$q.Promise}` and initialize `holdPromise_` to `null`.

File: geoportailv3/static/js/locationinfo/locationinfodirective.js [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+app.LocationinfoController.prototype.createFeatureOverlay_ =
+    function() {
+  var featureFill = new ol.style.Fill({
+    color: [255, 255, 0, 0.6]
+  });
+  var featureStroke = new ol.style.Stroke({
+    color: [255, 155, 55, 1],
+    width: 3
+  });
+  var featureOverlay = new ol.FeatureOverlay({
```
Comment (elemoine) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: `ol.FeatureOverlay` has been removed from OpenLayers 3, but we probably need to globally address that issue.

File: geoportailv3/static/js/maincontroller.js (hunk 16 above)
Comment (petzlux) [inline geoportailv3/static/js/maincontroller.js]: it keeps the information if there is a location info window being shown, or a layer information window being shown , or nothing. I think with strings, this is the clearest way

File: geoportailv3/static/js/locationinfo/locationinfodirective.js (hunk 20 above)
Comment (petzlux) [inline geoportailv3/static/js/locationinfo/locationinfodirective.js]: I tried that last week, but wouldnt pass compiler, as `this.$timeout_.cancel(this.holdPromise_)` expects `undefined`, not `null`:

````
ERR! compile /home/wepa/geoportailv3/geoportailv3/static/js/locationinfo/locationinfodirective.js:194: ERROR - actual parameter 1 of this.$timeout_.cancel does not match formal parameter
ERR! compile found   : (null|{catch: function ((function (?): ?|null)): ?, finally: function ((function (?): ?|null)): ?, then: function ((function (?): ?|null), (function (?): ?|null)=, (function (?): ?|null)=): ?})
ERR! compile required: (angular.$q.Promise|undefined)
ERR! compile           this.$timeout_.cancel(this.holdPromise_);
ERR! compile                                 ^
ERR! compile
ERR! compile /home/wepa/geoportailv3/geoportailv3/static/js/locationinfo/locationinfodirective.js:205: ERROR - actual parameter 1 of this.$timeout_.cancel does not match formal parameter
````

**Pull request merged**